// msw/lib/esm/rest-deps.js
import { R as RequestHandler, h as getUrlByMask, m as matchRequestUrl, g as getPublicUrlFromRequest, i as prepareRequest, j as prepareResponse, k as getTimestamp, l as getStatusCodeColor } from './RequestHandler-deps.js';
import { a as set, s as status, b as json, d as delay, f as fetch } from './fetch-deps.js';
import { c as cookie, b as body, t as text, x as xml } from './xml-deps.js';

/**
 * Performs a case-insensitive comparison of two given strings.
 */
function isStringEqual(actual, expected) {
    return actual.toLowerCase() === expected.toLowerCase();
}

var RESTMethods;
(function (RESTMethods) {
    RESTMethods["HEAD"] = "HEAD";
    RESTMethods["GET"] = "GET";
    RESTMethods["POST"] = "POST";
    RESTMethods["PUT"] = "PUT";
    RESTMethods["PATCH"] = "PATCH";
    RESTMethods["OPTIONS"] = "OPTIONS";
    RESTMethods["DELETE"] = "DELETE";
})(RESTMethods || (RESTMethods = {}));
const restContext = {
    set,
    status,
    cookie,
    body,
    text,
    json,
    xml,
    delay,
    fetch,
};
/**
 * Request handler for REST API requests.
 * Provides request matching based on method and URL.
 */
class RestHandler extends RequestHandler {
    constructor(method, mask, resolver) {
        super({
            info: {
                header: `${method} ${mask}`,
                mask,
                method,
            },
            ctx: restContext,
            resolver,
        });
        this.checkRedundantQueryParameters();
    }
    checkRedundantQueryParameters() {
        const { method, mask } = this.info;
        const resolvedMask = getUrlByMask(mask);
        if (resolvedMask instanceof URL && resolvedMask.search !== '') {
            const queryParams = [];
            resolvedMask.searchParams.forEach((_, paramName) => {
                queryParams.push(paramName);
            });
            console.warn(`\
[MSW] Found a redundant usage of query parameters in the request handler URL for "${method} ${mask}". Please match against a path instead, and access query parameters in the response resolver function:

rest.${method.toLowerCase()}("${resolvedMask.pathname}", (req, res, ctx) => {
  const query = req.url.searchParams
${queryParams
                .map((paramName) => `\
  const ${paramName} = query.get("${paramName}")`)
                .join('\n')}
})\
      `);
        }
    }
    parse(request) {
        return matchRequestUrl(request.url, this.info.mask);
    }
    getPublicRequest(request, parsedResult) {
        return Object.assign(Object.assign({}, request), { params: parsedResult.params });
    }
    predicate(request, parsedResult) {
        return (isStringEqual(this.info.method, request.method) && parsedResult.matches);
    }
    log(request, response) {
        const publicUrl = getPublicUrlFromRequest(request);
        const loggedRequest = prepareRequest(request);
        const loggedResponse = prepareResponse(response);
        console.groupCollapsed('[MSW] %s %s %s (%c%s%c)', getTimestamp(), request.method, publicUrl, `color:${getStatusCodeColor(response.status)}`, response.status, 'color:inherit');
        console.log('Request', loggedRequest);
        console.log('Handler:', {
            mask: this.info.mask,
            resolver: this.resolver,
        });
        console.log('Response', loggedResponse);
        console.groupEnd();
    }
}

function createRestHandler(method) {
    return (mask, resolver) => {
        return new RestHandler(method, mask, resolver);
    };
}
const rest = {
    head: createRestHandler(RESTMethods.HEAD),
    get: createRestHandler(RESTMethods.GET),
    post: createRestHandler(RESTMethods.POST),
    put: createRestHandler(RESTMethods.PUT),
    delete: createRestHandler(RESTMethods.DELETE),
    patch: createRestHandler(RESTMethods.PATCH),
    options: createRestHandler(RESTMethods.OPTIONS),
};

export { RestHandler as R, RESTMethods as a, restContext as b, isStringEqual as i, rest as r };
