// @ihollander/workshop-app/dist/react-app/exercise.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _reactRouterDom = require("react-router-dom");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _sandbox = _interopRequireDefault(require("./sandbox"));

var _fullpageLoader = _interopRequireDefault(require("./fullpage-loader"));

var _fourOhFour = _interopRequireDefault(require("./four-oh-four"));

var _theme = require("./theme");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

function Exercise(_ref) {
  var exercises = _ref.exercises;

  var _useParams = (0, _reactRouterDom.useParams)(),
      exerciseId = _useParams.exerciseId;

  var exerciseInfo = exercises[exerciseId];

  var _useState = (0, _react.useState)(true),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isReadmeVisible = _useState2[0],
      setIsReadmeVisible = _useState2[1];

  var _useState3 = (0, _react.useState)(true),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      isSandboxVisible = _useState4[0],
      setIsSandboxVisible = _useState4[1];

  var next = exercises[Number(exerciseId) + 1];
  var prev = exercises[Number(exerciseId) - 1];
  if (!exerciseInfo) return /*#__PURE__*/_react.default.createElement(_fourOhFour.default, null);

  var Readme = /*#__PURE__*/_react.default.lazy(exerciseInfo.readme.importFn);

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(ExerciseContainer, {
    split: isReadmeVisible && isSandboxVisible
  }, /*#__PURE__*/_react.default.createElement(ReadmeContainer, {
    isVisible: isReadmeVisible
  }, /*#__PURE__*/_react.default.createElement(ReadmeStyles, null, /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: /*#__PURE__*/_react.default.createElement(_fullpageLoader.default, null)
  }, /*#__PURE__*/_react.default.createElement(Readme, null), /*#__PURE__*/_react.default.createElement(ReadmeButtons, null, prev ? /*#__PURE__*/_react.default.createElement(_reactRouterDom.Link, {
    to: "/".concat(prev.readme.number)
  }, /*#__PURE__*/_react.default.createElement("span", {
    role: "img",
    "aria-labelledby": "arrow-left"
  }, "\u2190"), /*#__PURE__*/_react.default.createElement("span", null, "Previous")) : /*#__PURE__*/_react.default.createElement("div", null), next ? /*#__PURE__*/_react.default.createElement(_reactRouterDom.Link, {
    to: "/".concat(next.readme.number)
  }, /*#__PURE__*/_react.default.createElement("span", null, "Next"), /*#__PURE__*/_react.default.createElement("span", {
    role: "img",
    "aria-labelledby": "arrow-right"
  }, "\u2192")) : /*#__PURE__*/_react.default.createElement("div", null))))), /*#__PURE__*/_react.default.createElement(_sandbox.default, {
    exerciseInfo: exerciseInfo,
    isVisible: isSandboxVisible
  })), /*#__PURE__*/_react.default.createElement(WindowManager, null, /*#__PURE__*/_react.default.createElement("label", null, "Show Readme", " ", /*#__PURE__*/_react.default.createElement("input", {
    type: "checkbox",
    checked: isReadmeVisible,
    onChange: function onChange(e) {
      return setIsReadmeVisible(e.target.checked);
    }
  })), /*#__PURE__*/_react.default.createElement("label", null, "Show Exercise", " ", /*#__PURE__*/_react.default.createElement("input", {
    type: "checkbox",
    checked: isSandboxVisible,
    onChange: function onChange(e) {
      return setIsSandboxVisible(e.target.checked);
    }
  }))));
}

var WindowManager = _styledComponents.default.nav(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  position: fixed;\n  bottom: 0;\n  left: 0;\n  right: 0;\n  height: auto;\n  display: flex;\n  justify-content: flex-end;\n  gap: 1rem;\n  padding: 1rem;\n  background: var(--background-secondary-light);\n"])));

var ExerciseContainer = _styledComponents.default.main(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n  display: grid;\n  grid-template-columns: 100%;\n  grid-template-rows: ", ";\n  background: var(--background-primary);\n\n  @media (min-width: 800px) {\n    grid-template-columns: ", ";\n    grid-template-rows: 100%;\n  }\n"])), function (_ref2) {
  var split = _ref2.split;
  return split ? "50% 50%" : "100%";
}, function (_ref3) {
  var split = _ref3.split;
  return split ? "minmax(400px, 50%) 1fr" : "100%";
});

var ReadmeContainer = _styledComponents.default.div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2.default)(["\n  ", "\n  width: 100%;\n  max-width: 1000px;\n  margin: 0 auto;\n  overflow-y: auto;\n"])), function (_ref4) {
  var isVisible = _ref4.isVisible;
  return isVisible ? "" : "display: none;";
});

var ReadmeStyles = _styledComponents.default.section(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2.default)(["\n  padding: 2rem 2.5rem;\n  font-size: 1.2rem;\n\n  ", "\n\n  h1 {\n    margin: 0 -1rem;\n    background: var(--background-secondary);\n    border: 3px solid var(--color-primary);\n    box-shadow: 1px 1px var(--background-primary),\n      4px 4px var(--color-secondary);\n    border-radius: 3px;\n    padding: 1rem;\n    text-align: center;\n  }\n\n  a {\n    font-weight: var(--font-weight-bold);\n    transition: box-shadow 400ms ease 0s;\n    box-shadow: 0px 0px 0px var(--color-primary);\n  }\n\n  a:hover {\n    box-shadow: 0px 0px 0px var(--color-primary);\n    transition: box-shadow 100ms ease 0s;\n    box-shadow: 0px 2px 0px var(--color-primary);\n  }\n\n  img {\n    max-width: 100%;\n  }\n\n  > :not(pre) code {\n    display: inline;\n    font-family: Consolas, Monaco, \"Andale Mono\", \"Ubuntu Mono\", monospace;\n    font-size: 0.8em;\n    padding: 2px 6px;\n    margin: 1px;\n    border-radius: 3px;\n    background-color: var(--background-secondary-light);\n  }\n\n  > blockquote code {\n    background: var(--background-primary);\n  }\n\n  pre {\n    position: relative;\n    margin: 2rem -1rem;\n    background: var(--background-code);\n    border: 3px solid var(--color-primary);\n    box-shadow: 1px 1px var(--background-primary),\n      4px 4px var(--color-secondary);\n    border-radius: 3px;\n  }\n\n  pre code {\n    display: block;\n    overflow-y: auto;\n    padding: 1rem;\n  }\n\n  pre code[data-language]::before {\n    content: attr(data-language);\n    position: absolute;\n    top: -30px;\n    right: 10px;\n    font-size: 14px;\n    background: var(--color-primary);\n    padding: 0.25rem 0.5rem;\n    color: white;\n    border-top-left-radius: 3px;\n    border-top-right-radius: 3px;\n  }\n"])), _theme.prismTheme);

var ReadmeButtons = _styledComponents.default.div(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  justify-content: space-between;\n  padding: 2rem 0;\n\n  a {\n    display: inline-block;\n    max-width: 40%;\n    padding: 0.5rem;\n    margin: 0.5rem;\n    color: var(--font-color);\n    border: 3px solid var(--color-primary);\n    box-shadow: 4px 4px var(--color-secondary);\n    border-radius: 3px;\n    background: var(--background-secondary);\n    cursor: pointer;\n    outline: none;\n    transition: 100ms ease;\n  }\n\n  a:hover {\n    transform: translate(-2px, -2px);\n    box-shadow: 8px 8px var(--color-secondary);\n  }\n\n  a:focus {\n    border: 3px solid var(--color-secondary);\n    background-color: var(--color-secondary);\n  }\n\n  a span[role=\"img\"] {\n    margin: 0 0.5rem;\n    font-size: 1.25rem;\n  }\n"])));

var _default = Exercise;
exports.default = _default;
