// @ihollander/workshop-app/dist/react-app/header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _reactRouterDom = require("react-router-dom");

var _button = _interopRequireDefault(require("./styles/button"));

var _theme = require("./theme");

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

function Header(_ref) {
  var projectTitle = _ref.projectTitle,
      exercises = _ref.exercises;

  var _useTheme = (0, _theme.useTheme)(),
      colorMode = _useTheme.colorMode,
      setColorMode = _useTheme.setColorMode;

  var icon = colorMode === "light" ? "☀️" : "🌙";
  var numbers = exercises.filter(Boolean).map(function (ex) {
    return ex.readme.number;
  });
  return /*#__PURE__*/_react.default.createElement(Wrapper, null, /*#__PURE__*/_react.default.createElement(_reactRouterDom.Link, {
    to: "/"
  }, /*#__PURE__*/_react.default.createElement(TitleWrapper, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "logo"
  }, "//"), /*#__PURE__*/_react.default.createElement("h1", null, projectTitle), /*#__PURE__*/_react.default.createElement("h2", null, "Flatiron Workshops"))), /*#__PURE__*/_react.default.createElement(NavWrapper, null, /*#__PURE__*/_react.default.createElement(ExerciseNavWrapper, null, numbers.map(function (num) {
    return /*#__PURE__*/_react.default.createElement(_button.default, {
      key: num,
      as: _reactRouterDom.NavLink,
      to: "/".concat(num)
    }, num);
  })), /*#__PURE__*/_react.default.createElement(_button.default, {
    onClick: function onClick() {
      return setColorMode(colorMode === "light" ? "dark" : "light");
    }
  }, /*#__PURE__*/_react.default.createElement("span", {
    role: "img",
    "aria-label": colorMode
  }, icon))));
}

var Wrapper = _styledComponents.default.header(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  padding: 0 1rem;\n  background: var(--background-secondary-light);\n  border-bottom: 2px solid var(--background-primary);\n"])));

var TitleWrapper = _styledComponents.default.div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n  display: grid;\n  grid-template-areas:\n    \"logo title\"\n    \"logo subtitle\";\n  column-gap: 0.5rem;\n\n  .logo {\n    grid-area: logo;\n    font-weight: var(--font-weight-extra-bold);\n    color: var(--color-primary);\n    font-size: 2.25rem;\n    text-shadow: 1px 1px var(--background-secondary),\n      3px 3px var(--color-secondary);\n  }\n\n  h1 {\n    grid-area: title;\n    margin: 0;\n    padding: 0;\n    font-size: 1.25rem;\n  }\n\n  h2 {\n    grid-area: subtitle;\n    color: var(--font-color-light);\n    font-size: 1rem;\n    margin: 0;\n    padding: 0;\n  }\n"])));

var NavWrapper = _styledComponents.default.nav(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2.default)([""])));

var ExerciseNavWrapper = _styledComponents.default.div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2.default)(["\n  display: none;\n\n  @media (min-width: 600px) {\n    display: inline-block;\n  }\n"])));

var _default = Header;
exports.default = _default;
